### PR TITLE
MAINT: Do not use copyswap as part of indexing

### PR DESCRIPTION
**Pull request by seberg**
This is in an effort to just avoid copyswap for new DTypes.  Unfortunately, it does signficantly slow down copy of `object`.  It also slows down the copy of complex128, but I can special case that element size to make it faster than it was (in principle complex256 is probably still quite a lot slower, but it is niche).

Changes:
* `object` is about 33% slower for simple advanced indexing
* Structured dtypes are at least 50% faster for such cases.
* `complex128` is much faster, but this is due to a new fast-path.
* All other changes should be random fluctuations

Overall, I think I would be in favor of just going rolling with this, but also very interested if anyone has a good thought on a *followup* story below.

---

*Sorry for the story below... @ngoldbaum this is the continuation from your issue...*

The question is what alternatives exist.  I don't like `copyswap` (it is old-style with getting the array passed in, and doesn't really work for structured dtypes and more complicated parametric dtypes in general).

I could imagine a few ways to improve this:
1.  Maybe, we *should* have a simple "single item" copy function that is more lightweight.  If we do that that we probably have to _force_ users to provide this function for dtypes with references (without references, we can and should provide it).
   * As an alternative, we could limit it to *most* DTypes, e.g. by giving a signature of `DType->slots->get_item_copy_func(dtype)`, which would return a *lightweight* function with a signature of e.g. just `copy(dst, src, itemsize)`.  If your dtype is complicated enough that this doesn't work, don't use it.  One annoyance is that we would still need the current heavy-weight path *additionally* for structured dtypes or dtypes that fail to implement it.  OTOH, it could be a nice pattern overall.  Since we could write the "auto specialization" as:
   ```
   if (trivial_copy == &copy_8bit_func) {
       core_loop(trivial_copy, ...);
   }
   else if (trivial_copy != NULL {
       core_loop(trivial_copy, ...);  // same pattern
   }
   else if (trivial_copy == NULL) {
       core_loop(...);  // Version would use the heavy weight path as here.
   }
   ```
3. To really optimize it, doing something similar to the new `ufunc.at` approach has the potential of being 50% faster, rather than slower.
4. Just specializing the inner-loop for `N=1` (e.g. by abusing `strides=0`, since in that case `N` doesn't matter for copies at least), does recover quite a bit of the slowdown.

Overall, I am tempted to ignore the 1/3 slowdown for objects here.  In most code it is probably not very noticable (you are working with objects after all).  For a light-weight type with references (i.e. the new `StringDType`) the overhead might be larger

<details>  <summary> <b> Timing details </b> </summary>

Changes in (new) benchmarks:
```
       before           after         ratio
     [dd6a36c0]       [cf9989b4]
     <main>           <indexing-no-copyswap>
+        4.34±0μs       4.70±0.1μs     1.08  bench_indexing.IndexingWith1DArr.time_getitem_ordered((1000,), 'O')
+     6.31±0.09μs      6.83±0.07μs     1.08  bench_indexing.Indexing.time_op('complex128', 'indexes_', ':,I', '')
+      6.73±0.1μs      7.17±0.03μs     1.07  bench_indexing.Indexing.time_op('complex128', 'indexes_rand_', ':,I', '')
-     40.0±0.02μs      37.8±0.02μs     0.94  bench_indexing.Indexing.time_op('complex64', 'indexes_', 'np.ix_(I, I)', '=1')
-     7.20±0.01μs      6.79±0.06μs     0.94  bench_indexing.IndexingWith1DArr.time_setitem_ordered((2, 1000, 1), 'float16')
-     4.84±0.05μs      4.56±0.04μs     0.94  bench_indexing.Indexing.time_op('int16', 'indexes_rand_', ':,I', '=1')
-      33.5±0.5μs      31.6±0.01μs     0.94  bench_indexing.Indexing.time_op('float16', 'indexes_', 'np.ix_(I, I)', '')
-        6.15±0μs      5.73±0.03μs     0.93  bench_indexing.Indexing.time_op('int64', 'indexes_', ':,I', '')
-      40.6±0.2μs      37.8±0.02μs     0.93  bench_indexing.Indexing.time_op('complex64', 'indexes_rand_', 'np.ix_(I, I)', '=1')
-      34.8±0.3μs         32.4±1μs     0.93  bench_indexing.Indexing.time_op('int16', 'indexes_rand_', 'np.ix_(I, I)', '')
-     6.14±0.02μs      5.70±0.01μs     0.93  bench_indexing.Indexing.time_op('longdouble', 'indexes_', ':,I', '')
-     6.14±0.01μs      5.69±0.01μs     0.93  bench_indexing.Indexing.time_op('float64', 'indexes_', ':,I', '')
-     7.33±0.05μs      6.79±0.01μs     0.93  bench_indexing.IndexingWith1DArr.time_setitem_ordered((2, 1000, 1), 'int16')
-     5.97±0.01μs      5.52±0.01μs     0.92  bench_indexing.Indexing.time_op('longdouble', 'indexes_rand_', ':,I', '')
-     5.45±0.01μs      5.03±0.02μs     0.92  bench_indexing.Indexing.time_op('float16', 'indexes_rand_', ':,I', '')
-     5.98±0.02μs      5.50±0.01μs     0.92  bench_indexing.Indexing.time_op('int64', 'indexes_rand_', ':,I', '')
-     6.02±0.08μs      5.52±0.01μs     0.92  bench_indexing.Indexing.time_op('float64', 'indexes_rand_', ':,I', '')
-     5.53±0.01μs      5.07±0.04μs     0.92  bench_indexing.Indexing.time_op('int32', 'indexes_', ':,I', '')
-     7.65±0.01μs       7.01±0.1μs     0.92  bench_indexing.IndexingWith1DArr.time_setitem_ordered((1000, 2), 'complex64')
-     5.56±0.02μs      5.08±0.05μs     0.91  bench_indexing.Indexing.time_op('int32', 'indexes_rand_', ':,I', '')
-     1.96±0.02μs         1.79±0μs     0.91  bench_indexing.Indexing.time_op('float16', 'indexes_', 'I', '')
-     5.56±0.02μs      5.05±0.02μs     0.91  bench_indexing.Indexing.time_op('float32', 'indexes_rand_', ':,I', '')
-      5.69±0.1μs       5.15±0.2μs     0.91  bench_indexing.Indexing.time_op('int16', 'indexes_rand_', ':,I', '')
-     8.26±0.01μs      7.47±0.07μs     0.90  bench_indexing.IndexingWith1DArr.time_setitem_ordered((1000, 3), 'complex64')
-     5.65±0.06μs      5.06±0.06μs     0.90  bench_indexing.Indexing.time_op('float32', 'indexes_', ':,I', '')
-      5.81±0.1μs      5.03±0.02μs     0.87  bench_indexing.Indexing.time_op('int16', 'indexes_', ':,I', '')
-     5.92±0.08μs      5.04±0.01μs     0.85  bench_indexing.Indexing.time_op('float16', 'indexes_', ':,I', '')
-     7.61±0.01μs       6.27±0.1μs     0.82  bench_indexing.Indexing.time_op('complex64', 'indexes_', ':,I', '')
-     7.12±0.03μs      5.85±0.03μs     0.82  bench_indexing.IndexingWith1DArr.time_getitem_ordered((2, 1000, 1), 'complex64')
-        2.27±0μs         1.86±0μs     0.82  bench_indexing.IndexingWith1DArr.time_setitem_ordered((1000, 1), 'complex128')
-     7.55±0.01μs      5.97±0.01μs     0.79  bench_indexing.Indexing.time_op('complex64', 'indexes_rand_', ':,I', '')
-        2.18±0μs         1.65±0μs     0.75  bench_indexing.IndexingWith1DArr.time_getitem_ordered((1000, 1), 'complex128')
-     7.48±0.02μs      5.09±0.02μs     0.68  bench_indexing.Indexing.time_op('complex64', 'indexes_', 'I', '=1')
-     7.47±0.02μs      5.09±0.01μs     0.68  bench_indexing.Indexing.time_op('complex64', 'indexes_rand_', 'I', '=1')
-     1.47±0.01μs          978±7ns     0.66  bench_indexing.IndexingWith1DArr.time_getitem_ordered((1000,), 'complex128')
-        1.92±0μs         1.27±0μs     0.66  bench_indexing.IndexingWith1DArr.time_setitem_ordered((1000,), 'complex128')
-     25.5±0.05μs       12.6±0.2μs     0.50  bench_indexing.IndexingWith1DArr.time_getitem_ordered((1000,), 'i,O')
-     24.1±0.01μs      11.2±0.05μs     0.46  bench_indexing.IndexingWith1DArr.time_setitem_ordered((1000,), 'i,O')
```

Note that the benchmark sizes are relatively small.  The actual maximum slow-down is about 1/3 or 33% on my machine for `object` dtype.  `complex128` had a much heftier slowdown, so I added a special case for it's itemsize making it much faster instead.  Structured dtypes are much faster with this of course.

### CODE REVIEW

**Comment by ngoldbaum** [comment]
> Maybe, we should have a simple "single item" copy function that is more lightweight. If we do that that we probably have to force users to provide this function for dtypes with references (without references, we can and should provide it).

I don't think requiring this is a big deal. There's already a need to implement a single-element copy in the self to self cast, this just forces users to factor that out into a function that numpy itself can use when it needs single item buffer copies. The `.flat` iterator would be a customer for this API inside numpy (#23486). FWIW there also do seem to be a few users of copyswap outside scipy (albeit mostly in legacy user dtypes and wrappers for numpy in other languages): https://github.com/search?q=PyArray_CopySwapFunc&type=code

> As an alternative, we could limit it to most DTypes, e.g. by giving a signature of DType->slots->get_item_copy_func(dtype), which would return a lightweight function with a signature of e.g. just copy(dst, src, itemsize).

I'm fine with this API too. FWIW `stringdtype` is simple enough to implement this API.

> Overall, I am tempted to ignore the 1/3 slowdown for objects here. In most code it is probably not very noticable (you are working with objects after all). For a light-weight type with references (i.e. the new StringDType) the overhead might be larger

That's fine, we can always improve performance later using one of the approaches you outlined.

File: numpy/core/src/multiarray/lowlevel_strided_loops.c.src [hunk 1]
```diff
@@ -1528,7 +1533,13 @@ mapiter_trivial_@name@(PyArrayObject *self, PyArrayObject *ind,
             assert(npy_is_aligned(self_ptr, _UINT_ALIGN(@copytype@)));
             *(@copytype@ *)result_ptr = *(@copytype@ *)self_ptr;
 #else
-            copyswap(result_ptr, self_ptr, 0, self);
```
Comment (ngoldbaum) [inline numpy/core/src/multiarray/lowlevel_strided_loops.c.src:1531]: Out of curiosity, would leaving in an `object`-specific implementation that still does `copyswap` get back some of the performance loss? Or is it all in the overhead of setting up the `cast_info` struct?

**Review by ngoldbaum — APPROVED** [review]
Didn't spot any code issues, thanks for adding more benchmarks for this code path.

**Comment by ngoldbaum** [comment]
Also confirming that this fixes the seg fault I reported in https://github.com/numpy/numpy-user-dtypes/issues/64

**Comment by seberg** [comment]
Sounds like you would be +1 for a simple copy function (I would prefer that as a follow-up even if it may miss the next release)?

We could include swapping, but I really don't think swapping is a core API.  The thing that will push me there would be if I am sure that:
```
if (simple_copy_func == simple_copy8) {
    /* use simple_copy_func or simple_copy8 */
}
```
will work as a pattern for optimization.  What I am slightly unsure on is how easy this is, since we have to make sure the specialized code path is inlined.

**Comment by ngoldbaum** [comment]
> Sounds like you would be +1 for a simple copy function (I would prefer that as a follow-up even if it may miss the next release)?

Yeah, I don't need swapping for my use cases and I don't think many of the internal usages in numpy need it either.

**Comment by seberg** [comment]
Btw. thanks for showing the github code search.  Looks like it got a lot more useful!  Hadn't noticed that before.

**Comment by ngoldbaum** [comment]
I hit one of the other usages of `copyswap` in the indexing slow path and pushed a patch cribbed off this PR here: https://github.com/ngoldbaum/numpy/commit/7e6da5abf2994375cc86fbcf7781a6de5188d9c9. If it makes sense to you feel free to add it to this PR. I'm also happy to follow up after this is merged. I'm also happy to add more benchmarks, I wasn't sure if the existing ones already cover the fancy indexing slow path.

**Comment by seberg** [comment]
Oops, somehow thought the occurances there were changed before (which was is a bit silly in hindsight, the "trivial" version is very restrictive).  Pushed your commit here, thanks.

**Comment by ngoldbaum** [comment]
Not sure if the test failures are real. Looks like the mac test failure happened a few weeks ago on another PR: https://github.com/numpy/numpy/pull/23617#issuecomment-1515786287. Don't see references to the other failure in numpy's github history though...

**Comment by ngoldbaum** [comment]
I guess the other failure is https://github.com/numpy/numpy/issues/18069. Heisenbug yahtzee!

**Comment by ngoldbaum** [comment]
@mattip can I get you to look at this one too? It would be nice to have it in for 1.25.

**Comment by ngoldbaum** [comment]
Thanks for getting this merged!

> Any thoughts whether using this API would be faster than the current code?

Hopefully we’ll see speedups if we can write it so the dtypes can expose a single-element copy that compilers will inline. I’m not sure how hard that is.

**Pull request merged**